pkg/services/ngalert/state/state_tracker_test.go

```diff
@@ -145,64 +183,103 @@ func TestProcessEvalResults(t *testing.T) {
+					UID:     "test_uid",
+					OrgID:   123,
+					CacheId: "test_uid label1=value1, label2=value2",
+					Labels:  data.Labels{"label1": "value1", "label2": "value2"},
+					State:   eval.Normal,
+					Results: []StateEvaluation{
+						{evaluationTime, eval.Normal},
+						{EvaluationTime: evaluationTime.Add(1 * time.Minute), EvaluationState: eval.Normal},
+					},
+					StartsAt:           time.Time{},
+					EndsAt:             time.Time{},
+					LastEvaluationTime: evaluationTime.Add(1 * time.Minute),
 				},
 			},
 		},
 	}
 
 	for _, tc := range testCases {
-		t.Run("the correct number of entries are added to the cache", func(t *testing.T) {
+		t.Run("all fields for a cache entry are set correctly", func(t *testing.T) {
 			st := NewStateTracker(log.New("test_state_tracker"))
-			st.ProcessEvalResults(tc.uid, tc.evalResults, tc.condition)
-			assert.Equal(t, len(tc.expectedCacheEntries), len(st.stateCache.cacheMap))
+			_ = st.ProcessEvalResults(tc.uid, tc.evalResults, tc.condition)
+			for _, entry := range tc.expectedCacheEntries {
+				if !entry.Equals(st.Get(entry.CacheId)) {
+					t.Log(tc.desc)
+					printEntryDiff(entry, st.Get(entry.CacheId), t)
+				}
+				assert.True(t, entry.Equals(st.Get(entry.CacheId)))
+			}
 		})
 
-		t.Run("the correct state is set for each evaluation result", func(t *testing.T) {
+		t.Run("the expected number of entries are added to the cache", func(t *testing.T) {
 			st := NewStateTracker(log.New("test_state_tracker"))
 			st.ProcessEvalResults(tc.uid, tc.evalResults, tc.condition)
-			for _, entry := range tc.expectedCacheEntries {
-				testState := st.get(entry.CacheId)
-				assert.Equal(t, tc.expectedState, testState.State)
-			}
+			assert.Equal(t, len(tc.expectedCacheEntries), len(st.stateCache.cacheMap))
 		})
 
-		t.Run("the correct number of states are returned to the caller", func(t *testing.T) {
+		//This test, as configured, does not quite represent the behavior of the system.
+		//It is expected that each batch of evaluation results will have only one result
+		//for a unique set of labels.
+		t.Run("the expected number of states are returned to the caller", func(t *testing.T) {
 			st := NewStateTracker(log.New("test_state_tracker"))
 			results := st.ProcessEvalResults(tc.uid, tc.evalResults, tc.condition)
-			assert.Equal(t, tc.expectedReturnedStateCount, len(results))
+			assert.Equal(t, len(tc.evalResults), len(results))
 		})
+	}
+}
 
-		t.Run("the correct results are set for each cache entry", func(t *testing.T) {
-			st := NewStateTracker(log.New("test_state_tracker"))
-			_ = st.ProcessEvalResults(tc.uid, tc.evalResults, tc.condition)
-			for _, entry := range tc.expectedCacheEntries {
-				testState := st.get(entry.CacheId)
-				assert.Equal(t, len(entry.Results), len(testState.Results))
-				for i, res := range entry.Results {
-					assert.Equal(t, res, testState.Results[i])
-				}
-			}
-		})
+func printEntryDiff(a, b AlertState, t *testing.T) {
```

Comment:
This was added to help with test debugging

Reply:
Not important, but I have generally been happy with the output of go-cmp package (you can also pass it compare options to ignore certain properties). But I haven't looked at this output any maybe you are aware of that already - so grain of salt.

Reply:
I'll give it a look, thank you! I just got tired of staring are VERY similar but slightly different structs (and nested structs) so I decided to throw a quick function in here to quickly direct me to the differences.